cmd/gossamer/prune.go

```diff
@@ -0,0 +1,190 @@
+package main
+
+import (
+	"context"
+	"fmt"
+	"strings"
+
+	"github.com/ChainSafe/chaindb"
+	"github.com/ChainSafe/gossamer/dot/state"
+	"github.com/ChainSafe/gossamer/lib/blocktree"
+	"github.com/ChainSafe/gossamer/lib/common"
+	"github.com/ChainSafe/gossamer/lib/trie"
+	"github.com/dgraph-io/badger/v2"
+	"github.com/dgraph-io/badger/v2/pb"
+)
+
+//Pruner is an offline tool to prune the stale state with the
+// bloom filter, The workflow of pruner is very simple:
+// - iterate the storage state, reconstruct the relevant state tries
+// - iterate the database, stream all the targeted keys to new DB
+type Pruner struct {
+	inputDB        *chaindb.BadgerDB
+	storageState   *state.StorageState
+	blockState     *state.BlockState
+	bloom          *stateBloom
+	bestBlockHash  common.Hash
+	retainBlockNum int64
+
+	prunedDB *badger.DB
+}
+
+func newPruner(basePath string, bloomSize uint64, retainBlockNum int64) (*Pruner, error) {
+	db, err := loadChainDB(basePath)
+	if err != nil {
+		return nil, fmt.Errorf("failed to load DB %w", err)
+	}
+
+	base := state.NewBaseState(db)
+
+	bestHash, err := base.LoadBestBlockHash()
+	if err != nil {
+		return nil, fmt.Errorf("failed to get best block hash: %w", err)
+	}
+
+	// load blocktree
+	bt := blocktree.NewEmptyBlockTree(db)
+	if err = bt.Load(); err != nil {
+		return nil, fmt.Errorf("failed to load blocktree: %w", err)
+	}
+
+	// create blockState state
+	blockState, err := state.NewBlockState(db, bt)
+	if err != nil {
+		return nil, fmt.Errorf("failed to create block state: %w", err)
+	}
+
+	// create bloom filter
+	bloom, err := newStateBloomWithSize(bloomSize)
+	if err != nil {
+		return nil, fmt.Errorf("failed to create new bloom filter of size %d %w", bloomSize, err)
+	}
+
+	// load storage state
+	storageState, err := state.NewStorageState(db, blockState, trie.NewEmptyTrie())
+	if err != nil {
+		return nil, fmt.Errorf("failed to create new storage state %w", err)
+	}
+
+	return &Pruner{
+		inputDB:        db,
+		storageState:   storageState,
+		blockState:     blockState,
+		bloom:          bloom,
+		bestBlockHash:  bestHash,
+		retainBlockNum: retainBlockNum,
+	}, nil
+}
+
+func (p *Pruner) setBloomFilter() error {
+	// latest block header
+	header, err := p.blockState.GetHeader(p.bestBlockHash)
```

Comment:
ideally in the future we will keep unfinalized states in memory and only write them to the db if they get finalized

Reply:
Since we already leaving 256 recent blocks, wouldn't finalize block be a part of it?

Reply:
you mean unfinalized blocks? yes they should be, but in case someone sets the pruning to the most recent 6 blocks or something then it might prune un-finalized blocks